pkg/download/protocol.go

```diff
@@ -26,95 +34,121 @@ type Protocol interface {
 
 	// Zip implements GET /{module}/@v/{version}.zip
 	Zip(ctx context.Context, mod, ver string) (io.ReadCloser, error)
-
-	// Version is a helper method to get Info, GoMod, and Zip together.
-	Version(ctx context.Context, mod, ver string) (*storage.Version, error)
 }
 
-type protocol struct {
-	s  storage.Backend
-	dp Protocol
-	ch chan *job
-}
+// Wrapper helps extend the main stasher's functionality with addons.
+type Wrapper func(Protocol) Protocol
 
-type job struct {
-	mod, ver string
-	done     chan error
+// Opts specifies download protocol options to avoid long func signature.
+type Opts struct {
+	Storage   storage.Backend
+	Stasher   stash.Stasher
+	GoBinPath string
+	Fs        afero.Fs
 }
 
-// New takes an upstream Protocol and storage
-// it always prefers storage, otherwise it goes to upstream
-// and fills the storage with the results.
-func New(dp Protocol, s storage.Backend, workers int) Protocol {
-	ch := make(chan *job)
-	p := &protocol{dp: dp, s: s, ch: ch}
-	p.start(workers)
+// New returns a full implementation of the download.Protocol
+// that the proxy needs.
+func New(opts *Opts, wrappers ...Wrapper) Protocol {
```

Comment:
from this, it's not clear which wrapper of passed wrappers is outer most, 
decorators are one option, adding outer most info to docs another one